src/middleware/companyAuth.middleware.ts

```diff
@@ -0,0 +1,70 @@
+import ApplicationLogger from 'ch-logging/lib/ApplicationLogger'
+import { Session } from 'ch-node-session-handler'
+import { SessionKey } from 'ch-node-session-handler/lib/session/keys/SessionKey'
+import { SignInInfoKeys } from 'ch-node-session-handler/lib/session/keys/SignInInfoKeys'
+import { ISignInInfo } from 'ch-node-session-handler/lib/session/model/SessionInterfaces'
+import { NextFunction, Request, RequestHandler, Response } from 'express'
+
+import AuthConfig from 'app/models/authConfig'
+import { Mutable } from 'app/models/mutable'
+import Optional from 'app/models/optional'
+import DissolutionSession from 'app/models/session/dissolutionSession.model'
+import { JwtEncryptionService } from 'app/services/encryption/jwtEncryption.service'
+import SessionService from 'app/services/session/session.service'
+
+const OATH_SCOPE_PREFIX = 'https://api.companieshouse.gov.uk/company/'
+
+export default function CompanyAuthMiddleware(authConfig: AuthConfig,
+                                              encryptionService: JwtEncryptionService,
+                                              sessionService: SessionService,
+                                              logger: ApplicationLogger): RequestHandler {
+  return async (req: Request, res: Response, next: NextFunction) => {
+    const dissolutionSession: Optional<DissolutionSession> = sessionService.getDissolutionSession(req)
+
+    if (!dissolutionSession?.companyNumber) {
+      return next(new Error('No Company Number in session'))
+    }
+
+    const companyNumber = dissolutionSession!.companyNumber
+    const signInInfo: ISignInInfo = sessionService.getSignInInfo(req)
+
+    if (isAuthorisedForCompany(signInInfo, companyNumber)) {
+      logger.info(`User is authenticated for ${companyNumber}`)
+      return next()
+    }
+
+    logger.info(`User is not authenticated for ${companyNumber}, redirecting`)
+    return res.redirect(await getAuthRedirectUri(req, authConfig, encryptionService, sessionService, companyNumber))
+  }
+}
+
+function isAuthorisedForCompany(signInInfo: ISignInInfo, companyNumber: string): boolean {
+  return signInInfo[SignInInfoKeys.CompanyNumber] === companyNumber
+}
+
+async function getAuthRedirectUri(req: Request, authConfig: AuthConfig,
+                                  encryptionService: JwtEncryptionService,
+                                  sessionService: SessionService,
+                                  companyNumber?: string): Promise<string> {
+  const originalUrl: string = req.originalUrl
+  const scope: string = OATH_SCOPE_PREFIX + companyNumber
+  const nonce: string = encryptionService.generateNonce()
+  const encodedNonce: string = await encryptionService.jweEncodeWithNonce(originalUrl, nonce)
+
+  const mutableSession = sessionService.getSession(req) as Mutable<Session>
+  mutableSession.data[SessionKey.OAuth2Nonce] = nonce
+  sessionService.setSession(req, mutableSession as Session)
```

Comment:
This is fine, but just for context what I was suggesting was:
```
sessionService.setCompanyAuthNonce(nonce)
```
so that the middleware would just execute one line